austation/code/modules/mining/minebot_upgrades.dm

```diff
@@ -0,0 +1,269 @@
+
+
+/obj/item/minebot_upgrade  //  can't use mine_bot_upgrade because existing definitions conflict with what we want to do
+	name = "minebot upgrade"
+	icon_state = "door_electronics"
+	icon = 'icons/obj/module.dmi'
+	var/weight = 0  //  how many mod slots the upgrade takes up.  each minebot can only install mods until the total weight is 3
+	var/upgrade_id = null  //  populate with the DEFINES above, they point to the class of the upgrade
+	var/list/incompatible = list()  //  this list blacklists new mods from being added (such as gun upgrades for weapons you do not have equipped)
+
+/obj/item/minebot_upgrade/proc/is_minebotgun()
+	if(upgrade_id in GUNTYPES)
+		return TRUE
+
+/obj/item/minebot_upgrade/proc/is_minebotgun_upg()
+	if(upgrade_id in GUN_UPGRADES)
+		return TRUE
+
+/obj/item/minebot_upgrade/afterattack(mob/living/simple_animal/hostile/mining_drone/M, mob/user, proximity)
+	. = ..()
+	if(!istype(M) || !proximity || !upgrade_id)  //  something has gone very wrong.  the user probably wouldn't understand a warning for these
+		return
+	if(upgrade_id in M.upgrades_list)  //  can't install the same mod twice
+		to_chat(user, "<span class='warning'[M] already has a [src] installed!</span>")
+		return
+	if(M.mod_total >= 3  && !(upgrade_id in GUNTYPES))  //  we allow GUNTYPES because they replace an existing mod, instead of adding another.
+		to_chat(user, "<span class='warning'[M] has no more room to install a [src]!</span>")
+		return
+	for(var/obj/item/minebot_upgrade/U in M.upgrades_list)
+		if(upgrade_id in U.incompatible)
+			to_chat(user, "<span class='warning'there are conflicting mods already installed to \the [M]!  You may remove them with a crowbar.</span>")
+			return
+	src.upgrade_bot(M, user)
+
+/obj/item/minebot_upgrade/proc/upgrade_bot(mob/living/simple_animal/hostile/mining_drone/M, mob/user)
+	to_chat(user, "<span class='notice'you install a [src] upgrade into \the [M]</span>")
+	M.upgrades_list += upgrade_id
+	M.mod_total += src.weight
+	qdel(src)
+
+/obj/item/minebot_upgrade/proc/uninstall(mob/living/simple_animal/hostile/mining_drone/M, mob/user)
+	if(!get_turf(M))
+		if(is_minebotgun())
+			to_chat(user, "<span class='warning'there is no room to uninstall the [M]'s previous gun.</span>")
+		else
+			to_chat(user, "<span class='warning'there is no room to uninstall the [src].</span>")
+		return
+	var/turf/T = get_turf(M)
+	var/obj/item/gun/energy/E
+	if(M.minebot_gun)
+		E = M.minebot_gun
+
+	if(is_minebotgun())
+		for(var/obj/item/minebot_upgrade/gunmod in M.upgrades_list)
+			if(gunmod.is_minebotgun_upg())
+				gunmod.uninstall(M,user)
+		if(E)
+			to_chat(user, "<span class='notice'you uninstall [M]'s previous weapon kit, the [src].</span>")
+			qdel(E)
+
+	if(is_minebotgun_upg())
+		if(E)
+			to_chat(user, "<span class='notice'you uninstall the [src] from \the [M]'s [E]'.</span>")
+		else
+			to_chat(user, "<span class='notice'you removed pieces of a [src] from \the [M].</span>")
```

Comment:
```suggestion
			to_chat(user, "<span class='notice'>You uninstall [M]'s previous weapon kit, the [src].</span>")
			qdel(E)

	if(is_minebotgun_upg())
		if(E)
			to_chat(user, "<span class='notice'>You uninstall the [src] from \the [M]'s [E]'.</span>")
		else
			to_chat(user, "<span class='notice'>You removed pieces of a [src] from \the [M].</span>")
```

Reply:
Don't these automatically capitalise because they're at the start of a sentence?
I thought the standard practice was to avoid capitalising things except where they're nouns.
Guess I have a few changes to make to minebot.dm as well then.